Add WithOffset and OffsetParser

hobofan opened the pull request:
I added a `RawDocBlock` that also includes the offset positions of the original Markdown string.

I tried to make is as backwards compatible as possible, to preserve the simple interface of the original `DocBlock`.

One of the offset tests is ignored for now, since the original assumptions about parsing sections (https://github.com/hobofan/rust-docstrings/blob/7d7e31666f92cd614f0fdf479ca84ce8adf10cf2/src/extractors.rs#L57) don't quite hold and produce a offset that is incorrect by a few characters.

Commit:
Add WithOffset and OffsetParser

Wrap all the parsed parts in a WithOffset, preserving the offset in the
original Markdown string.

killercup commented on src/extractors.rs:
```diff
@@ -100,27 +108,27 @@ fn section<'a, I>(events: &mut I) -> Result<DocSection, ParseError> where
             DocSection::Custom(
                 headline,
-                md(events.take_while(not!(start Tag::Header(1))))),
+                md(events.map(|(event, _)| event).take_while(not!(start Tag::Header(1))))),
```
Okay, I've read that `map(|(event, _)| event)` code enough times! :D Can you add a `TupelIterExt` trait that is implemented for all `Iterator<Item=(T1, T2)>` which has `fn fst(&self) -> T1` and `fn snd(&self) -> T2`?

hobofan replied:
Do you have some pointers for that? I tried some things, but the closest I got was something like `.unzip().0`, which consumes the whole iterator, which isn't really something we want here.

killercup replied:
I'm currently sitting at the train station in Cologne and it's -3°C, so I only wrote this: https://play.rust-lang.org/?gist=cdabc6c2cac0cd069ec7522bf136983b&version=stable&backtrace=0

hobofan replied:
Oh, I was too focused on the trait part. That looks good, applying it now.

killercup replied:
Yeah, well, implementing a custom map specialization is nice, but probably much more work and not that much shorter to type in the end.

killercup commented on src/extractors.rs:
```diff
@@ -19,32 +19,39 @@ macro_rules! not {
+        } else {
+            0
+        }
+    };
```
Is that extra block necessary to drop the result of events.peek() earlier? If you implement the `fst` and `snd` fns I mentioned this could become `events.peek().snd().unwrap_or(0)` ;)

killercup commented on src/offset_parser.rs:
```diff
@@ -0,0 +1,22 @@
+                }
+                None => return None
+            }
+        }
```
I don't get why this is in a `loop`, a branches return

killercup commented on src/types.rs:
```diff
@@ -21,6 +21,18 @@ pub struct DocBlock {
+    pub description: Option<(Documentation, usize)>,
+    /// Sections
+    pub sections: Vec<(DocSection, usize)>,
+}
```
I have yet to read all the code, but one thing I noticed: Why'd you only add the start byte offset? I'd probably add a byte range instead of usize. Or even a custom `struct Span {start: Position, end. Position}` with `struct Position { byte_offset, line, col}` and then a wrapper `struct WithSpan<T> { inner: T, span: Span }` which `impl<Inner> Deref for WithSpan<Inner> { type Target = T; … }`.

Also, I'm fine with breaking changes as there are no production users I know of and this is 0.1. Let's combine this with DocBlock :)

killercup commented on src/offset_parser.rs:
```diff
@@ -0,0 +1,22 @@
+    fn next(&mut self) -> Option<(Event<'a>, usize)> {
+        loop {
+            let &mut OffsetParser(ref mut inner_iter) = self;
+            let offset = inner_iter.get_offset();
```
Re <https://github.com/killercup/rust-docstrings/pull/1/files#r96317886>: Oh, I see: pulldown only gives us a usize. Any chance we can get the size here to calculate the whole span?

killercup reviewed (commented):
Nice! What use case are you thinking of? Writing lints with correct spans?

If've not read all the code, but it's good so far. I left some suggestions inline. Let me know what you think :)

hobofan commented:
Those are all godd points, I'll see what I can do.

We might be able to build a Span with the End Events emitted by the parser, though I'm not sure if they are emitted early enough to be useful (or a few lines after the block actually ends), but lets try and see.

hobofan commented:
> Nice! What use case are you thinking of? Writing lints with correct spans?

You know it! :)

hobofan commented:
I rewrote the tuples into `WithOffset`s and removed the RawDocBlock again.

Looking into it, constructing a Span with and end only seems to be possible with a larger rewrite from the simpler Iterator version to a more complex tree walking version. 

For my usecase simple offsets would be enough, so I would leave it at that for now, if that's okay with you.

killercup commented on src/offset_parser.rs:
```diff
@@ -0,0 +1,20 @@
+
+        match inner_iter.next() {
+            Some(event) => {
+                Some((event, offset))
```
Any reason not to return `Some(WithOffset::new(event, offset))` here (and let that change trickle down)? If you make its fields public you can destructure it like a tuple. (Or make it a tuple struct?)

killercup reviewed (commented):
Nice, thanks for the quick update!

Do you want to add more tests to make coveralls happy? I don't really care as long as it's >90%.

killercup commented:
ping @hobofan

Do you want to continue to add to this PR? Otherwise I'll merge it as-is.

hobofan commented:
@killercup I still want to try the thing you said in the last comment. I should be able to get to it this weekend.

hobofan commented:
@killercup Sorry, I don't think that I will find time to finish the PR anytime soon. :/

killercup commented:
No worries! I hope to find some time to work on this a bit more to get it into clippy soon, and I'll probably use this branch :)
